### ISSUE TITLE
Bug: <Memory Leak persist after update 1.13.1 >

### ISSUE THREAD

**Issue by mchinchilla** [mikependon/RepoDb#1142]
### Bug Description

Hi Michael,
I'm using RepoDb.PostgreSql.BulkOperations version 1.13.1, this application parse a lot of CDR's files and insert in a PostgreSql database, but when my application execute this line:

```csharp
await db.InsertAllAsync( tableName: "schema.table", recordsFinal, batchSize: insertBuffer, transaction: transaction );
```

Memory usage increase and never decrease, No matter if I force to GC to collect and clean.
This application runs in a Debian Linux Server with 32 GB and postgresql 13.x, I'm using the latest version of RepoDb.PostgreSql.BulkOperations and dotnet core 6.0.
When the applications usage gets the 32 GB of RAM, the server starts to swap and the application stops working.

And it seems that it is incremental, for example it starts with 100 MB and for each file it processes it increases exponentially 200MB, 400 MB... 3 GB, etc.

**Images**
![image](https://user-images.githubusercontent.com/5103578/225972551-9702874d-4602-4f6d-b041-85128a1470a1.png)

**Library Version:**
Version: RepoDb.PostgreSql.BulkOperations version 1.13.1

**Comment by mikependon** [comment]
For more information, I am referencing these [discussions](https://twitter.com/mike_pendon/status/1636489405098491910) from Twitter.
Link: https://twitter.com/mike_pendon/status/1636489405098491910

**Comment by mikependon** [comment]
I have investigated this and it sounds to me that it works as expected. But, you as a user must know the caveat of it. **In short, this is not a memory leaks!**

RepoDB requires these caching for it to perform a more performant insertion of the batch operations.

### Explanation

Below is the screenshot and a very small project that we used for simulation and replication. The project below requires a SQL Server as we did the simulation there. Unfortunately for SQL Server and MDS, it only allow a maximum number of 2100 parameters. Therefore, you will see that I can only hit the max of `20 batchSize` for `20 rows` with `50 columns data entities`.

**Project:**

This project is good for small simulation on this kind of issue. 
[InsertAllMemoryLeaks-InsertAll.zip](https://github.com/mikependon/RepoDB/files/11006610/InsertAllMemoryLeaks-InsertAll.zip)

**Screenshot:**

![image](https://user-images.githubusercontent.com/591015/226058366-d7b5511e-154b-4296-a03a-9239182855ca.png)

### What the program does?

- It first create a table named `[dbo].[BigTable]` with 120 columns.
- The program will iterate infinitely to do an `InsertAll` operation.
- In every iteration, the program will create a list of `BigTable` entity with max 50 columns. The creation of the list will vary on the maximum batchSize, from 1. This is to ensure that RepoDB will create and cache each buffer's command text in the memory.
- Insert the created list of `BigTable` entity towards the table.

### Observation

**Fluctuations:**

In the first few seconds, the memory has fluctuated a lot, it is because when the `INSERT` statement is being created for the number of rows given, the library will put that in the cache.

Truth: If you insert 1 row, it will create 1 `INSERT` statement and cache it. If you insert 2 rows, it will create 2 `INSERT` statements and cache it, and so forth.

The `batchSize` is the maximum number of `INSERT` statement it will create and cache into the memory. So in short, you will have 20 times `INSERT` statement being cached into the memory (in which each of them will have different number of parameters based on the columns provided on the data entities).

**Flat-Lines:**

You will notice the flat-lines after those 20 `INSERT` statement has been cached to the memory. This is because the library is not creating an `INSERT` statement anymore, instead, simply reusing the one in the cache based on the number of rows you are inserting.

### Behavior Extent

This kind of behavior is expected and is also present to both `MergeAll` and `UpdateAll`.

### Conclusion

The number of cached statements will vary on the number of `batchSize` you passed on the batch operations (i.e.: `InsertAll`, `MergeAll` and `UpdateAll`). The size of the statement that is being cached will vary on the size of the entity schema (i.e.: Number of Columns).

### Optimizations

Currently, RepoDB is creating multiple INSERT statement per row-numbers's batch insertion. See below.

- InsertAll with 1 row:
   Statement cached:
   ```
   INSERT INTO [Table] VALUES (ColumnFirst, ..., ColumnLast);
   ```
- InsertAll with 5 rows:
  Statements cached:
   ```
  INSERT INTO [Table] VALUES (ColumnFirst, ..., ColumnLast);
  INSERT INTO [Table] VALUES (ColumnFirst2, ..., ColumnLast2);
  INSERT INTO [Table] VALUES (ColumnFirst3, ..., ColumnLast3);
  INSERT INTO [Table] VALUES (ColumnFirst4, ..., ColumnLast4);
  INSERT INTO [Table] VALUES (ColumnFirst5, ..., ColumnLast5);
   ```
- And so forth...

The statement above are verbose and is also not using the more optimal bulk insert. This can be optimized by below.

- InsertAll with 3 rows:
  Statements cached:
   ```
  INSERT INTO [Table]
  VALUES
  (ColumnFirst, ..., ColumnLast),
  (ColumnFirst2, ..., ColumnLast2),
  (ColumnFirst3, ..., ColumnLast3);
   ```

With that approach, it will eliminate so may characters from the memory.

**Comment by mikependon** [comment]
Referencing: https://github.com/mikependon/RepoDB/issues/380

**Comment by mikependon** [comment]
We will create the simulation in the PGSQ database, we are hopeful that this 2100 limit is not present there so we can simulate your use-case. We will post the result here once done.

**Comment by mikependon** [comment]
If you use `BulkInsert` (equivalent to `BinaryBulkImport`), you are levaraging the real bulk operations and these cached are skipped. In the screenshot below, you will notice that we adjusted the `batchSize` to `1000` with `100` columns and yet the memory usage is not increasing.

![image](https://user-images.githubusercontent.com/591015/226057314-b5f2f73f-9767-4b14-bc13-7c98902b37a8.png)

You will also notice that the higher the `batchSize` the faster it inserts the data. It is because the real bulk insert loves bring all the data at once from the client into the database server - there it performs much better.

Note: If you bulk insert 1000 rows and you set the batchSize to 1, it will iterate 1000 times (this is the behavior of SqlBulkCopy itself). Therefore, if you bulk insert 1000 rows and you set the batchSize to 1000 as well, it will bring all data at once.

EDIT: Attaching. [InsertAllMemoryLeaks-BulkInsert.zip](https://github.com/mikependon/RepoDB/files/11006586/InsertAllMemoryLeaks-BulkInsert.zip)

**Comment by mikependon** [comment]
Hmmm - interestingly, seems I can replicate your issue in PGSQL. I modified the program and enable the 100 max batch size with 120 columns in a table.

Below is a screenshot in the first few seconds of run. It first burst up to 900 MB, went down suddenly and slowly climbing up, exactly as what you explained in Twitter.

![image](https://user-images.githubusercontent.com/591015/226060089-cdcddf43-8b5a-4836-b835-cf7fd26c6f52.png)

And few minutes after, the memory is not going down. It even reach to 2 GB, and is still climbing.

![image](https://user-images.githubusercontent.com/591015/226060392-43cbc7a6-09a2-42a2-984b-7fbae3c33d5d.png)

But, when the library had cached all the row batch statements (100 total statements), the execution suddenly becomes faster and it gives a flat line on the memory allocations.

![image](https://user-images.githubusercontent.com/591015/226060703-c8a2f9e0-5b6d-42a9-85a4-e724d0fcb81c.png)

After 5 minutes - the line is now flat and is not climbing anymore.

![image](https://user-images.githubusercontent.com/591015/226060789-e272bbbc-7e31-4197-b3c2-28739ed992af.png)

Here is the project that replicates your case. 
[InsertAllMemoryLeaksPostgreSql.zip](https://github.com/mikependon/RepoDB/files/11006675/InsertAllMemoryLeaksPostgreSql.zip)

**In conclusion, this is not a memory leak.** Since you're requiring a big caching on the statement based on the big data entity you have, the library requires such memory to execute your insertion fast.

**Comment by mikependon** [comment]
The `BinaryBulkInsert` only requires 40 MB as it does not cache anything.

![image](https://user-images.githubusercontent.com/591015/226061325-db2a358e-6d12-4338-b1dc-a7ce2bdbce62.png)

Project: [InsertAllMemoryLeaksPostgreSql-BinaryBulkInsert.zip](https://github.com/mikependon/RepoDB/files/11006687/InsertAllMemoryLeaksPostgreSql-BinaryBulkInsert.zip)

**Comment by rhuijben** [comment]
My fork on https://github.com/ampscm/repodb uses a smarter trick than caching just every batch size. It uses a set of smartly chosen batchsizes so usuallly you don't have to bother any more. It picks 1,2,3,4,5,6,7,8,9,10 like the old version and then 20, 50, 100, 200, 500, 1000, 2000 upto the max supported with the columns/provider. So mostly like how you would use banknotes after the initial steps.

This appears to have a much better behavior in the common user scenarios. (Max double ram from before, but proper bulk inserts for really large sets)
